Episode 1. Introduction to AI coding assistants (#11)
olgaminaeva/gen-ai-coding: episodes/Introduction-AI-coding.md

---
title: "Introduction to AI coding assistants"
teaching: 20
exercises: 10
---

:::::::::::::::::::::::::::::::::::::: questions 

- How do AI coding assistants work?
- What are the main AI coding assistants and what are their main characteristics?
- How to set up Codeium as coding assistant for the lesson?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Learn about AI models behind coding assistants
- Outline the main AI coding assistants and their capabilities
- Set up Codeium as coding assistant for the lesson

::::::::::::::::::::::::::::::::::::::::::::::::

## Introduction to AI coding assistants ##

> AI models for coding assistants, such as Codeium and GitHub Copilot, rely on machine learning (ML) techniques, particularly deep learning and natural language processing (NLP), to assist developers. These models are trained on vast amounts of publicly available code and documentation to understand patterns, syntax, and code logic across various programming languages. 

Reviewer: ```suggestion
Artificial intelligence (AI) models for coding assistants, such as Codeium and GitHub Copilot, rely on machine learning (ML) techniques, particularly deep learning and natural language processing (NLP), to assist developers. These models are trained on vast amounts of publicly available code and documentation to understand patterns, syntax, and code logic across various programming languages. 
```

Let’s break down the *key characteristics* of how these models work:

| Characteristic | Description | Examples |
| --- | ----------- | ------- |

> | Code Understanding through Machine Learning | AI coding assistants are built on models that analyze and learn from vast amounts of data, including open-source codebases, libraries, and developer behavior. These models break down code into a mathematical representation that the AI can interpret. They learn common patterns, best practices, syntax rules, and how developers approach specific tasks or problems. | For example, when you start typing, the AI assistant predicts what you are likely to write next based on patterns it has learned. These predictions can be simple line completions or even more complex, multi-line code suggestions.|

Reviewer: ```suggestion
| Code Understanding through Machine Learning | AI coding assistants are built on models that analyze and learn from vast amounts of data, including open-source codebases, libraries, and developer behavior. These models break down code into a mathematical representation that the AI can interpret. They learn common patterns, best practices, syntax rules, and how developers approach specific tasks or problems. | When you start typing, the AI assistant predicts what you are likely to write next based on patterns it has learned. These predictions can be simple line completions or even more complex, multi-line code suggestions.|
```
The col is called "Examples", so no need to repeat that it's an example

> | Code Understanding through Machine Learning | AI coding assistants are built on models that analyze and learn from vast amounts of data, including open-source codebases, libraries, and developer behavior. These models break down code into a mathematical representation that the AI can interpret. They learn common patterns, best practices, syntax rules, and how developers approach specific tasks or problems. | For example, when you start typing, the AI assistant predicts what you are likely to write next based on patterns it has learned. These predictions can be simple line completions or even more complex, multi-line code suggestions.|

Reviewer: ```suggestion
| Code understanding through ML | AI coding assistants are built on models that analyze and learn from vast amounts of data, including open-source codebases, libraries, and developer behavior. These models break down code into a mathematical representation that the AI can interpret. They learn common patterns, best practices, syntax rules, and how developers approach specific tasks or problems. | For example, when you start typing, the AI assistant predicts what you are likely to write next based on patterns it has learned. These predictions can be simple line completions or even more complex, multi-line code suggestions.|
```

Maybe do a check about the acronyms: the first time you use them, specify them fully, but afterwards make sure you're using the acronym directly

> | Context Awareness | AI coding assistants are designed to understand context at multiple levels, including code, variables, and functions. This makes them able to suggest relevant code suggestions based on the context. | For example, if you're writing a loop to iterate over a list of items, the AI can suggest the entire loop structure based on what it recognizes in the surrounding code.|

Reviewer: ```suggestion
| Context Awareness | AI coding assistants are designed to understand context at multiple levels, including code, variables, and functions. This makes them able to suggest relevant code suggestions based on the context. | If you're writing a loop to iterate over a list of items, the AI can suggest the entire loop structure based on what it recognizes in the surrounding code.|
```

> | Context Awareness | AI coding assistants are designed to understand context at multiple levels, including code, variables, and functions. This makes them able to suggest relevant code suggestions based on the context. | For example, if you're writing a loop to iterate over a list of items, the AI can suggest the entire loop structure based on what it recognizes in the surrounding code.|

Reviewer: ```suggestion
| Context awareness | AI coding assistants are designed to understand context at multiple levels, including code, variables, and functions. This makes them able to suggest relevant code suggestions based on the context. | For example, if you're writing a loop to iterate over a list of items, the AI can suggest the entire loop structure based on what it recognizes in the surrounding code.|
```

Capitalizing titles is fine, but make sure to be coherent across all titles then (up to your preference whether to do it or not)

> | Natural Language Processing (NLP) | AI models are both trained to understand code syntax and to interpret natural language. This is crucial for features like chat-based interaction or when a developer types a comment or command in plain English, expecting the AI to generate code. | If you write a comment like `"Create a function to fetch user data from an API"`, the AI can generate the function code based on that request. You can also ask questions, like `"How do I format dates in Python?"`, and the AI will provide an answer or relevant code snippet.

Reviewer: You've already defined NLP, so you can just use the acronym here

> | Natural Language Processing (NLP) | AI models are both trained to understand code syntax and to interpret natural language. This is crucial for features like chat-based interaction or when a developer types a comment or command in plain English, expecting the AI to generate code. | If you write a comment like `"Create a function to fetch user data from an API"`, the AI can generate the function code based on that request. You can also ask questions, like `"How do I format dates in Python?"`, and the AI will provide an answer or relevant code snippet.

Reviewer: ```suggestion
| NLP | AI models are both trained to understand code syntax and to interpret natural language. This is crucial for features like chat-based interaction or when a developer types a comment or command in plain English, expecting the AI to generate code. | If you write a comment like `"Create a function to fetch user data from an API"`, the AI can generate the function code based on that request. You can also ask questions, like `"How do I format dates in Python?"`, and the AI will provide an answer or relevant code snippet.
```

> | Learning from user interaction | AI coding assistants also learn and improve from ongoing user interactions. Over time, they adjust to individual developer preferences and coding styles, offering more personalized and relevant suggestions | For example, as users accept or reject suggestions, the assistant refines its future outputs based on this feedback (so-called "feedback loop"). Many AI coding assistants can learn from private codebases (when privacy policies allow) to better align with specific project structures, libraries, or functions common to a particular team or organization (custom adaptations). |

Reviewer: See comment above, here the words in the title are not capitalized ("Learning from user interaction"), while in other titles (e.g., "Code Generation and Refactoring") of the table are. Pick one way and adopt it in the all table

With a foundation of how AI models for coding assistants function — how they analyze code, offer suggestions, and improve efficiency — let’s explore some of the most popular AI coding assistants available today. Each tool uses AI in slightly different ways to enhance your coding experience. In the following table, we'll compare these tools to give you an overview of their key features, pricing, and capabilities. This will help you understand where Codeium fits and what alternatives you might consider.

## Popular AI coding assistants ##

> | **Feature** | **GitHub Copilot** | **Codeium** | **Amazon CodeWhisperer / Amazon Q Developer** | **Tabnine** | **Hugging Chat**  |

Reviewer: ```suggestion
|   | **GitHub Copilot** | **Codeium** | **Amazon CodeWhisperer / Amazon Q Developer** | **Tabnine** | **Hugging Chat**  |
```

|------------|-----------|-----------------|-----------------|-------------------|------------------|
|  | ![](episodes/fig/github copilot logo.png) | ![](episodes/fig/codeium_logo.png) | ![](episodes/fig/amazon Q developer logo.png) | ![](episodes/fig/Tabnine-AI logo.png) | ![](episodes/fig/hugging chat logo.webp) |
| **Brief Introduction**  | AI-powered coding assistant developed in collaboration with OpenAI, suggests code as you type within the IDE. | AI coding assistant that enhances productivity with intelligent code completions and refactoring suggestions. | AI assistant from Amazon Web Services, offers real-time code recommendations and integrates into various development environments. | AI-powered code completion tool providing accurate, context-aware code suggestions across multiple languages. | Leverages open-source models for flexible and customizable AI code assistance. |

> | **Key Features** | – Integration with GitHub ecosystem<br>– Multilingual support<br>– Acts as a virtual pair programmer  | – Advanced code completions<br>– Code refactoring capabilities<br>– Supports multiple programming languages | – Code recommendations<br>– Security scans for vulnerabilities<br>– Real-time documentation assistance<br>– Multilingual support  | – Robust autocompletion<br>– Learns from codebase to suggest relevant snippets and APIs<br>– Extensive language support  | – Open-source model integration<br>– High flexibility in deployment<br>– Supports diverse coding environments |

Reviewer: There is something that is not working with the listed/bulleted points here: 
![image](https://github.com/user-attachments/assets/e2b450f2-e04a-4219-9891-a42ef401d329)

Now that we've compared some of the top AI coding assistants, you can see that each tool offers unique features and benefits. For this lesson, we’ll focus on teaching you how to use AI coding assistants with Codeium as our primary example. Codeium offers a powerful, beginner-friendly experience that helps you speed up coding through intelligent completions, refactoring suggestions, and support for multiple programming languages.

In the next section, we’ll walk through the steps to set up Codeium in your development environment, so you can start taking advantage of these features right away.

::::::::::::::::::::::::::::::::::::: challenge 

> ## Setting up Codeium ##

Reviewer: I'd make this a separate regular section (like ### Using Codeium as your coding assistant), and not a challenge. Challenges special sections are meant for exercises :)
